File: 643-maximum-average-subarray-i/maximum-average-subarray-i.cpp
class Solution {
public:
    double findMaxAverage(vector<int>& nums, int k) {
        int i=0, j=0;
        double maxavg = INT_MIN, sum = 0;
        while (j<nums.size()){
            sum+=nums[j];
            if (j-i+1 < k){
                j+=1;
            } else if(j-i+1 == k){
                double avg = sum/k;
                maxavg = max(maxavg, avg);
                j+=1;
                sum-=nums[i];
                i+=1;
            } else {
                while (j-i+1 > k){
                    sum-=nums[i];
                    i+=1;
                }
                if (j-i+1 == k){
                    double avg = sum/k;
                    maxavg = max(maxavg, avg);
                }

                j+=1;
            }
        }
        return maxavg;
    }
};

File: 643-maximum-average-subarray-i/maximum-average-subarray-i.go
func findMaxAverage(nums []int, k int) float64 {
    var max_avg float64 = -math.MaxFloat64
    var sum int
    j := 0
    for i := 0; i < len(nums); i++ {
        sum += nums[i]
        win_len := i - j + 1
        
        for win_len > k {
            sum -= nums[j]
            j++
            win_len = i - j + 1
        }

        if win_len == k {
            avg := float64(sum) / float64(k)
            max_avg = max(max_avg, avg)
        }
    }

    return max_avg
}